compressed-headers.md

```diff
@@ -0,0 +1,147 @@
+| 629,474          | 629,474     | 0              | 100
+
+### Service Bit
+
+A new service bit is required so that the nodes can advertise their ability to supply compact headers. Dash uses bit 11 and designates it `NODE_HEADERS_COMPRESSED`.
+
+### P2P Messages
+
+Three new messages would be used by nodes that enable compact block header support, two query messages: `getheaders2` and `sendheaders2` and one response: `headers2`.
+
+#### `getheaders2` -- Requesting compact headers
+
+The new p2p message required to request compact block headers would require the same fields as the current `getheaders` message:
+
+|Field Size | Description          | Data type | Comments
+| -         | -                    | -         | -
+|4          |version              |uint32_t  |the protocol version
+|1+         |hash count           |var_int   |number of block locator hash entries
+|32+        |block locator hashes |char[32]  |block locator object; newest back to genesis block (dense to start, but then sparse)
+|32         |hash_stop            |char[32]  |hash of the last desired block header; set to zero to get as many blocks as possible (2000)
+
+#### `sendheaders2` -- Request compact header announcements
+
+Since [BIP-130](https://github.com/bitcoin/bips/blob/master/bip-0130.mediawiki), nodes have been able to request to receive new headers directly in `headers` messages, rather than via an `inv` of the new block hash and subsequent `getheader` request and `headers` response (followed by a final `getdata` to get the tip block itself, if desired). This is requested by transmitting an empty `sendheaders` message after the version handshake is complete.]
+
+Upon receipt of this message, the node is permitted, but not required, to preemptively announce new headers with the `headers2` message (instead of `inv`). Preemptive header announcement has been supported by the protocol version ≥ 70206 | Dash Core version ≥ 0.12.1.
+
+For the motivational use-case it makes sense to also update this mechanism to support sending header updates using compact headers using a new message.
+
+#### `headers2` -- Receiving compact headers
+
+A `headers2` message is returned in response to `getheaders2` or at new header announcement following a `sendheaders2` request. It contains both `length` and `headers` fields. The `headers` field contains a variable length vector of `block_header2`:
+
+| Field Size | Description | Data type       | Comments
+| -          | -           | -               | -
+|1+         |length      |var_int         |Length of `headers`
+|39-81x?    |headers     |block_header2[] |Compressed block headers in [block_header2 data type](#block_header2-data-type) format
+
+### Implementation
+
+* The first header in the first `block_header2[]` vector to a newly-connected client MUST contain the full `nBits`, `timestamp`, `version` and `prev_block_hash` fields, along with a correctly populated `bitfield` byte.
```

Comment:
Not only the first header to a newly-connected client, but the first header of each `block_header2` vector.  Otherwise, some data races may appear when deserializing the `block_header2` vector (explanation: https://github.com/dashpay/dash/pull/4497#issue-1021695959)

Reply:
Thanks @gabriel-bjg . Does 700b0e3f97102d060d6e4dccf53d54b1ae67be48 address this sufficiently?